bindings/rust/src/bytes.rs

```diff
@@ -31,15 +34,18 @@ impl Deref for Bytes {
 }
 
 impl Bytes {
-    pub fn slice(&self, start: usize, end: usize) -> Self {
-        Self::from(&self.0[start..end])
-    }
-
-    pub fn slice_from(&self, start: usize) -> Self {
-        self.slice(start, self.len())
-    }
-
-    pub fn slice_to(&self, end: usize) -> Self {
-        self.slice(0, end)
+    pub fn slice(&self, range: impl RangeBounds<usize>) -> Self {
+        let len = self.len();
+        let begin = match range.start_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(&n) => n + 1,
+            Bound::Excluded(&n) => n,
+            Bound::Unbounded => len,
+        };
+        Self::from(&self.0[begin..end])
```

Comment:
Looks ugly. Maybe std or bytes crate has already done this somewhere, and we can just forward it to `Vec<u8>`.

Reply:
-  This piece of code is only used in `#[no_std]`.
- This piece of code is copied from [bytes](https://github.com/tokio-rs/bytes/blob/729bc7c2084a42fda2c62da6933951fa7ac875aa/src/bytes.rs#L205-L246).
  Since there is no thread in `#[no_std]`, we don't have to use `bytes`.
  [And `Vec<u8>` is faster and binary is smaller.](https://github.com/nervosnetwork/molecule/pull/13#issue-356435816)
  So I use a new type of `Vec<u8>` in `#[no_std]`, and keep the same function signature as `bytes`.
- As present, [`::core::slice::SliceIndex`] is not able to be implemented by users.
  Here are lots of magics, in [`::core::slice::SliceIndex`].
  I didn't find a way to define a type `T`:
  ```rust
  fn func(vec: &Vec<u8>, range: T) -> &[u8] {
      &vec[range]
  }
  ```
- No methods use `RangeBounds` to slice a `Vec<u8>` or `&[u8]`.

[`::core::slice::SliceIndex`]: https://doc.rust-lang.org/core/slice/trait.SliceIndex.html